### PR TITLE
Workspace system, CI & Accessibility 

### PR DESCRIPTION
**Pull request by MathisBullinger**
Please only merge after the [user service pull request](https://github.com/eutopia-platform/service_user/pull/4) has been approved and deployed.

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
   name: 'App',
+  errorCaptured(err) {
+    this.handleError(err)
+    return false
+  },
+  methods: {
+    handleError(err) {
+      if (
+        err.message.startsWith('Network error') ||
+        (err.graphQLERrors &&
```
Comment (jeyj0) [inline src/App.vue]: I think there's a typo here? `err.graphQLERrors` instead of `err.graphQLErrors`

File: src/components/atomic/Alert.vue [hunk 2]
```diff
@@ -0,0 +1,51 @@
+<template>
+  <div :class="getClass" @click="close">{{ msg }}</div>
```
Comment (jeyj0) [inline src/components/atomic/Alert.vue:2]: The alert div should get `role="alert"` for accessibility. (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role)

File: src/components/atomic/Loader.vue [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<template>
+  <div :class="getClass"></div>
```
Comment (jeyj0) [inline src/components/atomic/Loader.vue]: You might want to add `role="progressbar"` and `aria-label="loading"` :)

File: src/components/molecular/Confirmation.vue [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<template>
+  <div class="getClass">
```
Comment (jeyj0) [inline src/components/molecular/Confirmation.vue]: shouldn't this be `:class="getClass"`?

File: src/components/molecular/LabeledInput.vue [hunk 5]
```diff
@@ -0,0 +1,58 @@
+<template>
+  <div :class="getClass">
+    <Paragraph>{{ label }}:</Paragraph>
+    <Input
+      small
+      :size="size"
+      :placeholder="placeholder"
+      :value="defaultValue"
+      @input="onInput"
+    />
```
Comment (jeyj0) [inline src/components/molecular/LabeledInput.vue:11]: Since a label is available here, why not set `aria-label` on the Input? :thinking: 
Or use html-native `<label>` instead of / wrapping the `<Paragraph>` or its contents.

File: src/components/molecular/LoginForm.vue [hunk 6]
```diff
@@ -1,22 +1,18 @@
 <template>
   <div :class="getClass">
```
Comment (jeyj0) [inline src/components/molecular/LoginForm.vue]: Can we use an actual `<form>` here?

File: src/components/molecular/LoginForm.vue [hunk 7]
```diff
@@ -34,6 +30,26 @@ export default new Molecular({
   components: {
     Popup,
   },
+  created() {
+    localStorage.removeItem('sessionToken')
```
Comment (jeyj0) [inline src/components/molecular/LoginForm.vue]: It might be worth extracting this into a separate file that can be called whenever necessary, and to clarify what is happening here (resetting the browser to avoid data leaks when not logged in).

File: src/components/molecular/UserStatus.vue [hunk 8]
```diff
@@ -46,6 +42,18 @@ export default {
         })
         .then(() => {
           localStorage.removeItem('sessionToken')
+          localStorage.removeItem('workspace')
```
Comment (jeyj0) [inline src/components/molecular/UserStatus.vue]: (partially) same code as in `LoginForm`. That proves the worth to extract it into a separate script file :smile:

File: src/pages/LoadingScreen.vue [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<template>
+  <div>
```
Comment (jeyj0) [inline src/pages/LoadingScreen.vue:2]: Is the wrapping `<div>` actually necessary? Seems to be quite irrelevant... (trying to keep the amount of elements in the browser low)

File: src/pages/invite/InviteError.vue [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            name
+            callname
+            email
+            id
+          }
+        }
+      `,
+      error(err) {
+        if (
+          err.message.replace('GraphQL error:', '').trim() === 'NOT_LOGGED_IN'
```
Comment (jeyj0) [inline src/pages/invite/InviteError.vue:36]: I've seen most of this code before... Can it be moved to the custom Error class? :thinking:

File: src/pages/invite/InviteError.vue [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      email: '',
+    },
+  },
+  methods: {
+    logout: function() {
+      this.$apollo.provider.clients.auth
+        .mutate({
+          mutation: gql`
+          mutation {
+            logout(token: "${localStorage.getItem('sessionToken')}")
```
Comment (jeyj0) [inline src/pages/invite/InviteError.vue]: How about using GraphQL variables here? :)

File: src/pages/invite/InviteError.vue [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+          }
+        `,
+        })
+        .then(() => {
+          localStorage.removeItem('sessionToken')
```
Comment (jeyj0) [inline src/pages/invite/InviteError.vue]: This is again the same code as in two other places, as I've noted there.

File: src/pages/workspace/Settings.vue [hunk 13]
```diff
@@ -1,11 +1,200 @@
 <template>
-  <div>
-    Settings
+  <div class="settings">
+    <Header tertiary>Settings</Header>
+    <div class="content">
+      <Card>
+        <div>
```
Comment (jeyj0) [inline src/pages/workspace/Settings.vue]: Is the div inside the Card necessary?

File: src/pages/workspace/Settings.vue [hunk 14]
```diff
@@ -1,11 +1,200 @@
...
+    confirmDelete: false,
+  },
+  methods: {
+    submitProfile: function() {
+      this.$apollo.mutate({
+        mutation: gql`
+          mutation { 
+            setName(name: "${this.user.name}" callname: "${
+          this.user.callname
+        }") {
```
Comment (jeyj0) [inline src/pages/workspace/Settings.vue:130]: Please use GraphQL variables here, and maybe re-do the formatting :)

File: src/pages/workspace/WorkspaceIndicator.vue [hunk 15]
```diff
@@ -0,0 +1,32 @@
+<template>
+  <div :class="getClass">
```
Comment (jeyj0) [inline src/pages/workspace/WorkspaceIndicator.vue]: is this wrapping div necessary?

File: src/pages/workspace/settings/InviteForm.vue [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+              break
+            case 'INVALID_EMAIL':
+              this.$emit('error', new CubeError('invalid email'))
+              break
+            case 'ALREADY_MEMBER':
+              this.$emit('error', new CubeError('already member of workspace'))
+              break
+            default:
+              this.$emit('error', new CubeError())
+              throw err
```
Comment (jeyj0) [inline src/pages/workspace/settings/InviteForm.vue:46]: What do you think of this code?
```
const previousMsg = err.message.replace(...).trim()
const msg = {
  ALREADY_INVITED: 'already invited to workspace',
  INVALID_EMAIL: 'invalid email'
  ALREADY_MEMBER: 'already member of workspace'
}[previousMsg]
this.$emit('error', new CubeError(msg))
if (msg !== undefined) throw err
```

I think it does the same, but much more extendable... :thinking: Haven't tested it though.

**Review by jeyj0 — CHANGES_REQUESTED** [review]
I added a few suggestions and a few things that need changing (duplicate code). :smile:

File: src/App.vue (hunk 1 above)
Comment (MathisBullinger) [reply src/App.vue]: fixed with 5f8c89c

File: src/components/atomic/Alert.vue (hunk 2 above)
Comment (MathisBullinger) [reply src/components/atomic/Alert.vue:2]: added role with cfbe90f

File: src/components/atomic/Loader.vue (hunk 3 above)
Comment (MathisBullinger) [reply src/components/atomic/Loader.vue]: Now has `aria-label="loading"` (bca12f4). `role="progressbar"` would be misleading because it's not a progress bar — it doesn't have a defined start, end or progress.

File: src/components/molecular/Confirmation.vue (hunk 4 above)
Comment (MathisBullinger) [reply src/components/molecular/Confirmation.vue]: fixed with 69b3358

File: src/components/molecular/LabeledInput.vue (hunk 5 above)
Comment (MathisBullinger) [reply src/components/molecular/LabeledInput.vue:11]: The component now uses an html label and binds the label to the input 73d1734

File: src/pages/LoadingScreen.vue (hunk 9 above)
Comment (MathisBullinger) [reply src/pages/LoadingScreen.vue:2]: This one is necessary. The loader rotates and we don't want the entire page to rotate — because so would every element you add to it. Otherwise [this](https://eutopiaworkspace.slack.com/files/UFPRQC11S/FJ9NB5VBL/screen_recording_2019-05-12_at_16.17.03.mov) happens.

File: src/pages/workspace/Settings.vue (hunk 14 above)
Comment (MathisBullinger) [reply src/pages/workspace/Settings.vue:130]: done 7808796

File: src/pages/workspace/Settings.vue (hunk 13 above)
Comment (MathisBullinger) [reply src/pages/workspace/Settings.vue]: removed 699dbdf

File: src/pages/workspace/WorkspaceIndicator.vue (hunk 15 above)
Comment (MathisBullinger) [reply src/pages/workspace/WorkspaceIndicator.vue]: removed 699dbdf

File: src/components/molecular/LoginForm.vue (hunk 6 above)
Comment (MathisBullinger) [reply src/components/molecular/LoginForm.vue]: yes we can 555a888

File: src/components/molecular/LoginForm.vue (hunk 7 above)
Comment (MathisBullinger) [reply src/components/molecular/LoginForm.vue]: done bffe2d1

File: src/components/molecular/UserStatus.vue (hunk 8 above)
Comment (MathisBullinger) [reply src/components/molecular/UserStatus.vue]: done bffe2d1

File: src/pages/invite/InviteError.vue (hunk 12 above)
Comment (MathisBullinger) [reply src/pages/invite/InviteError.vue]: done bffe2d1

File: src/pages/invite/InviteError.vue (hunk 11 above)
Comment (MathisBullinger) [reply src/pages/invite/InviteError.vue]: done 7808796

**Comment by MathisBullinger** [comment]
I would like to rethink how the errors are handled in the frontend and implement a proper system for that at a later point. Because of that, I don't think it is very important to work on the error message parsing now — it will be changed anyway.

File: src/pages/workspace/Settings.vue [hunk 17]
```diff
@@ -3,58 +3,52 @@
     <Header tertiary>Settings</Header>
     <div class="content">
```
Comment (jeyj0) [inline src/pages/workspace/Settings.vue:4]: is a `<main>` tag appropriate? Since the class is "content", I assume it is.

File: src/components/molecular/LoginForm.vue [hunk 18]
```diff
@@ -14,11 +14,12 @@
       Submit
```
Comment (jeyj0) [inline src/components/molecular/LoginForm.vue:14]: Is it possible to give this button `type="submit"`? That would make the form more accessible. Be careful though, it'll also make it behave differently, I think (if it isn't already). You'll probably have to catch the native "submit" event, so the form doesn't reload the page when pressing Enter or the submit button.

**Review by jeyj0 — DISMISSED** [review]
I added a few more thoughts. You can merge before or after looking into them. :smile:

**Review by jeyj0 — DISMISSED** [review]

**Review by jeyj0 — APPROVED** [review]
I think theres room for generalization here. I think an atomic or molecular component might make sense here. It would then capture the default event and emit a submit event itself. Ive approved the changes though, because I dont think its necessary to do that in this PR
